File: src/client/HttpClient.ts
import { SDKConfig, BASE_URLS } from "./Config";
import { SDKError } from "../errors/SDKError";
import { AuthError } from "../errors/AuthError";
import { NetworkError } from "../errors/NetworkError";
import { ApiResponse } from "../types/common";

/**
 * HTTP client aligned with Adhere backend contract.
 *
 * Two auth modes:
 *  - request()         → Authorization: Bearer <apiKey>        (session creation)
 *  - sessionRequest()  → x-access-token: <sessionToken>        (all other endpoints)
 *  - uploadWithSession() → multipart + x-access-token          (file uploads)
 */
export class HttpClient {
  private baseUrl: string;
  private apiKey: string;
  private timeout: number;
  private sessionToken: string | null = null;

  constructor(config: SDKConfig) {
    this.baseUrl = BASE_URLS[config.environment || "sandbox"];
    this.apiKey = config.apiKey;
    this.timeout = config.timeout || 30000;
  }

  setSessionToken(token: string): void {
    this.sessionToken = token;
  }

  /**
   * API-key authenticated request (used for session creation only).
   * Header: Authorization: Bearer <apiKey>
   */
  async request<T>(
    method: string,
    path: string,
    body?: unknown
  ): Promise<T> {
    return this._fetch<T>(path, {
      method,
      headers: {
        "Authorization": `Bearer ${this.apiKey}`,
        "Content-Type": "application/json",
      },
      body: body ? JSON.stringify(body) : undefined,
    });
  }

  /**
   * Session-token authenticated request (used for all endpoints after session creation).
   * Header: x-access-token: <sessionToken>
   */
  async sessionRequest<T>(
    method: string,
    path: string,
    body?: unknown
  ): Promise<T> {
    this.requireToken();

    return this._fetch<T>(path, {
      method,
      headers: {
        "x-access-token": this.sessionToken!,
        "Content-Type": "application/json",
      },
      body: body ? JSON.stringify(body) : undefined,
    });
  }

  /**
   * Multipart file upload with session-token auth.
   * Header: x-access-token: <sessionToken>
   * NOTE: Do NOT set Content-Type — browser sets it with boundary automatically.
   */
  async uploadWithSession<T>(path: string, formData: FormData): Promise<T> {
    this.requireToken();

    return this._fetch<T>(path, {
      method: "POST",
      headers: {
        "x-access-token": this.sessionToken!,
      },
      body: formData,
    });
  }

  /**
   * Legacy upload with API key (kept for backward compatibility if needed).
   */
  async upload<T>(path: string, formData: FormData): Promise<T> {
    return this._fetch<T>(path, {
      method: "POST",
      headers: {
        "Authorization": `Bearer ${this.apiKey}`,
      },
      body: formData,
    });
  }

  // ---- Internal ----

  private requireToken(): void {
    if (!this.sessionToken) {
      throw new AuthError(
        "No session token set. Call sdk.createSession() first.",
        "NO_SESSION_TOKEN"
      );
    }
  }

  private async _fetch<T>(path: string, init: RequestInit): Promise<T> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(`${this.baseUrl}${path}`, {
        ...init,
        signal: controller.signal,
      });

      const contentType = response.headers.get("content-type") || "";
      if (!contentType.includes("application/json")) {
        throw new SDKError(
          `Unexpected response from ${path} (${response.status})`,
          response.status
        );
      }

      // Parse the backend's standard response envelope
      const envelope: ApiResponse = await response.json();

      if (!response.ok || envelope.status === "error") {
        const statusCode = response.status;

        // Map specific HTTP statuses to error types
        if (statusCode === 401 || statusCode === 403) {
          throw new AuthError(
            envelope.message || "Authentication failed",
            envelope.code
          );
        }

        throw new SDKError(
          envelope.message || `Request to ${path} failed`,
          statusCode,
          envelope.code,
          envelope.data as Record<string, unknown>
        );
      }

      // Return the unwrapped data from the envelope
      return envelope.data as T;

    } catch (err: any) {
      if (err instanceof SDKError) throw err;

      if (err.name === "AbortError") {
        throw new NetworkError("Request timeout");
      }
      if (err.name === "TypeError" && err.message?.includes("fetch")) {
        throw new NetworkError(`Network error: ${err.message}`);
      }

      throw new SDKError(
        err.message || "Unknown error",
        0,
        "UNKNOWN_ERROR"
      );
    } finally {
      clearTimeout(timer);
    }
  }
}

File: flutter/lib/src/client/http_client.dart
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;

import '../errors/sdk_error.dart';
import 'config.dart';

/// HTTP client for the SmartComply SDK.
///
/// Auth strategy:
/// - [createSession] is called with `Authorization: Bearer <apiKey>`
/// - All subsequent calls use `Authorization: Bearer <sessionToken>`
///   Once a session token is set via [setSessionToken], it takes precedence.
class HttpClient {
  final String _baseUrl;
  final String _apiKey;
  final Duration _timeout;
  String? _sessionToken;

  HttpClient(SDKConfig config)
      : _baseUrl = kBaseUrls[config.environment]!,
        _apiKey = config.apiKey,
        _timeout = config.timeout;

  /// Called by SmartComply after a successful session creation.
  void setSessionToken(String token) => _sessionToken = token;

  /// Clears the session token (e.g. after session revocation).
  void clearSessionToken() => _sessionToken = null;

  String get _bearerToken => _sessionToken ?? _apiKey;



  /// Sends a JSON request. Use [useApiKey] = true to force api_key auth
  /// (used only for session creation before a token exists).
  Future<T> request<T>(
    String method,
    String path, {
    Map<String, dynamic>? body,
    bool useApiKey = false,
    required T Function(Map<String, dynamic>) fromJson,
  }) async {
    final uri = Uri.parse('$_baseUrl$path');
    final authHeader = useApiKey ? _apiKey : _bearerToken;
    final headers = {
      'Authorization': 'Bearer $authHeader',
      'Content-Type': 'application/json',
    };

    late http.Response response;
    try {
      switch (method.toUpperCase()) {
        case 'GET':
          response = await http.get(uri, headers: headers).timeout(_timeout);
          break;
        case 'POST':
          response = await http
              .post(uri, headers: headers,
                  body: body != null ? jsonEncode(body) : null)
              .timeout(_timeout);
          break;
        default:
          throw SDKError('Unsupported HTTP method: $method');
      }
    } on SocketException {
      throw SDKError('Network unavailable', 0);
    } on http.ClientException catch (e) {
      throw SDKError('Network error: ${e.message}', 0);
    }

    return _parseResponse(response, path, fromJson);
  }

  /// Multipart upload — used for liveness/create and liveness/submit.
  Future<T> uploadMultipart<T>(
    String path, {
    required Map<String, String> fields,
    required List<http.MultipartFile> files,
    required T Function(Map<String, dynamic>) fromJson,
  }) async {
    final uri = Uri.parse('$_baseUrl$path');

    final multipart = http.MultipartRequest('POST', uri)
      ..headers['Authorization'] = 'Bearer $_bearerToken'
      ..fields.addAll(fields)
      ..files.addAll(files);

    late http.StreamedResponse streamed;
    try {
      streamed = await multipart.send().timeout(_timeout);
    } on SocketException {
      throw SDKError('Network unavailable', 0);
    }

    final response = await http.Response.fromStream(streamed);
    return _parseResponse(response, path, fromJson);
  }

  T _parseResponse<T>(
    http.Response response,
    String path,
    T Function(Map<String, dynamic>) fromJson,
  ) {
    final contentType = response.headers['content-type'] ?? '';
    if (!contentType.contains('application/json')) {
      throw SDKError(
        'Unexpected response from $path (${response.statusCode})',
        response.statusCode,
      );
    }

    final data = jsonDecode(response.body) as Map<String, dynamic>;

    if (response.statusCode < 200 || response.statusCode >= 300) {
      final msg = data['detail'] ??
          data['message'] ??
          data['error'] ??
          jsonEncode(data);
      throw SDKError('$path failed: $msg', response.statusCode);
    }

    return fromJson(data);
  }
}
